jax/interpreters/masking.py

```diff
@@ -340,23 +361,74 @@ def lift(self, val):
     return MaskTracer(self, val, onp.shape(val))
 
   def sublift(self, val):
-    return MaskTracer(self, val.val, val.shape_expr)
+    return MaskTracer(self, val.val, val.polymorphic_shape)
 
   def process_primitive(self, primitive, tracers, params):
-    vals, shape_exprs = unzip2((t.val, t.shape_expr) for t in tracers)
+    vals, polymorphic_shapes = unzip2((t.val, t.polymorphic_shape) for t in tracers)
     if primitive in shape_parameterized_primitive_rules:
       rule = shape_parameterized_primitive_rules[primitive]
-      out, out_shape = rule(shape_envs, vals, shape_exprs, **params)
+      out, out_shape = rule(shape_envs, vals, polymorphic_shapes, **params)
     else:
       avals = [t.aval for t in tracers]
       out = primitive.abstract_eval(*avals, **params)
       out_shape = [o.shape for o in out] if primitive.multiple_results else out.shape
-      logical_shapes = map(partial(eval_polymorphic_shape, values_dict=shape_envs.logical), shape_exprs)
-      out = masking_rules[primitive](vals, logical_shapes, **params)
+      logical_shapes = map(shape_as_value, polymorphic_shapes)
+      masking_rule = masking_rules.get(primitive)
+      if masking_rule is None:
+        raise NotImplementedError('Masking rule for {} not implemented yet.'.format(primitive))
+      out = masking_rule(vals, logical_shapes, **params)
     if not primitive.multiple_results:
       return MaskTracer(self, out, out_shape)
     else:
       return map(partial(MaskTracer, self), out, out_shape)
 
-  def process_call(self, call_primitive, f: lu.WrappedFun, tracers, params):
-    raise NotImplementedError  # TODO mask-of-jit
+  def process_call(self, call_primitive, f, tracers, params):
+    vals, polymorphic_shapes = unzip2((t.val, t.polymorphic_shape) for t in tracers)
+    f, out_shapes_thunk = mask_subtrace(f, self.master, polymorphic_shapes)
```

Comment:
We need to be careful about the signature of `mask_subtrace`. The jit cache is keyed on transformations and the 'static' arguments provided to them, in this case `master` and `polymorphic_shapes`, and currently the jitted function will have no way of seeing changes to the logical shapes.

You can see the problem if you do:
```python
In [1]: import jax
   ...: import numpy as np
   ...: import functools

In [2]: @functools.partial(jax.mask, in_shapes=['n'], out_shape='')
   ...: def padded_sum(x):
   ...:   return jax.jit(np.sum)(x)
   ...: 
   ...: print(padded_sum([np.array([3, 1, 4, 1, 5])], dict(n=3)))  # 8
8

In [3]: print(padded_sum([np.array([3, 1, 4, 1, 5])], dict(n=2)))  # Should be 4
8
```
I'm not sure how to fix this right now, just pointing out that it's incorrect 😁.

Reply:
I think to get this working we need to move the `extend_shape_envs` call from `masking.mask_fun` into `masking.mask_subtrace`. We also need to
1. Move `padded_env` binding into `mask_subtrace` (from api.mask)
2. Pass `logical_env` (in flattened form) to mask_subtrace as a dynamic argument

I'm wondering whether that should be in a separate pr though?

Reply:
Resolved.